### ISSUE TITLE
Invalid SQL generation.

### ISSUE THREAD

**Issue by powerumc** [PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#1300]
### Steps to reproduce
Create Table
```
date DATETIME DEFAULT 'CURRENT_TIMESTAMP' PRIMARY KEY,
uid INT NOT NULL PRIMARY KEY
```

C# Code
```cs
var item = new Item {
    Uid = 1
};
dbContext.SaveChangesAsync(cancellationToken);
```

Then occurred `DbUpdateConcurrencyException` exception because it is invalid SQL.

Generated SQL
```sql
INSERT INTO `item` (`date`, `uid`)
VALUES (@p0, @p1);
SELECT `date`
FROM `item`
WHERE ROW_COUNT() = 1 AND `uid` = @p0 AND `date` = LAST_INSERT_ID();
```

`` `date` = LAST_INSERT_ID() `` is wrong.

### The issue
Describe what is not working as expected.

```
Exception message: Microsoft.EntityFrameworkCore.DbUpdateConcurrencyException: Database operation expected to affect 1 row(s) but actually affected 0 row(s). Data may have been modified or deleted since entities were loaded. See http://go.microsoft.com/fwlink/?LinkId=527962 for information on understanding and handling optimistic concurrency exceptions.

Stack trace:
Microsoft.EntityFrameworkCore.DbUpdateConcurrencyException: Database operation expected to affect 1 row(s) but actually affected 0 row(s). Data may have been modified or deleted since entities were loaded. See http://go.microsoft.com/fwlink/?LinkId=527962 for information on understanding and handling optimistic concurrency exceptions.
   at Microsoft.EntityFrameworkCore.Update.AffectedCountModificationCommandBatch.ThrowAggregateUpdateConcurrencyException(Int32 commandIndex, Int32 expectedRowsAffected, Int32 rowsAffected)
   at Microsoft.EntityFrameworkCore.Update.AffectedCountModificationCommandBatch.ConsumeResultSetWithPropagationAsync(Int32 commandIndex, RelationalDataReader reader, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Update.AffectedCountModificationCommandBatch.ConsumeAsync(RelationalDataReader reader, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Update.ReaderModificationCommandBatch.ExecuteAsync(IRelationalConnection connection, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Update.ReaderModificationCommandBatch.ExecuteAsync(IRelationalConnection connection, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.ExecuteAsync(IEnumerable`1 commandBatches, IRelationalConnection connection, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.ExecuteAsync(IEnumerable`1 commandBatches, IRelationalConnection connection, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChangesAsync(IList`1 entriesToSave, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChangesAsync(DbContext _, Boolean acceptAllChangesOnSuccess, CancellationToken cancellationToken)
   at Pomelo.EntityFrameworkCore.MySql.Storage.Internal.MySqlExecutionStrategy.ExecuteAsync[TState,TResult](TState state, Func`4 operation, Func`4 verifySucceeded, CancellationToken cancellationToken)
```

### Further technical details

MySQL version: 8.0.19
Operating system: Windows 10
Pomelo.EntityFrameworkCore.MySql version: 3.2.0
Microsoft.AspNetCore.App version: 3.1.0

**Comment by mguinness** [comment]
Your create table definition results in `Multiple primary key defined`, please provide valid DDL.

Also you should await async operations, i.e. `await dbContext.SaveChangesAsync(cancellationToken);`

**Comment by powerumc** [comment]
Sorry for the short question.

I made the source code for you to test.
I use docker MySQL.

Step 1. Create MySQL Server.
```bash
docker run -it --name mysql -p 3306:3306 -e 'DB_REMOTE_ROOT_NAME=root' -e 'DB_REMOTE_ROOT_PASS=test' -e 'DB_NAME=testdb' sameersbn/mysql:latest
```

Step 2. I made a Project File and a single file Console Program.
```xml
<Project Sdk="Microsoft.NET.Sdk">

    <PropertyGroup>
        <OutputType>Exe</OutputType>
        <TargetFramework>net5.0</TargetFramework>
    </PropertyGroup>

    <ItemGroup>
      <PackageReference Include="Microsoft.EntityFrameworkCore" Version="3.1.8" />
      <PackageReference Include="Microsoft.EntityFrameworkCore.Design" Version="3.1.8">
        <PrivateAssets>all</PrivateAssets>
        <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
      </PackageReference>
      <PackageReference Include="Microsoft.Extensions.Logging.Console" Version="5.0.0" />
      <PackageReference Include="Pomelo.EntityFrameworkCore.MySql" Version="3.2.4" />
    </ItemGroup>

</Project>
```

```cs
using System;
using System.ComponentModel.DataAnnotations.Schema;
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Design;
using Microsoft.Extensions.Logging;

namespace TestPomelo
{
    class Program
    {
        static void Main(string[] args)
        {
            var dbContext = new DbContextFactory().CreateDbContext();

            // It's OK
            // var test = new Test
            // {
            //     Date = DateTime.Now
            // };

            // But Invalid SQL;
            // INSERT INTO `test`
            // VALUES ();
            // SELECT `uid`, `date`
            // FROM `test`
            // WHERE ROW_COUNT() = 1 AND `uid` = LAST_INSERT_ID() AND `date` = LAST_INSERT_ID();
            var test = new Test
            {
            };
            dbContext.Tests.Add(test);
            dbContext.SaveChanges();
        }
    }

    public class DbContextFactory : IDesignTimeDbContextFactory<TestDbContext>
    {
        public TestDbContext CreateDbContext(params string[] args)
        {
            var optionBuilder = new DbContextOptionsBuilder<TestDbContext>();
            optionBuilder.UseMySql("Server=localhost;Database=testdb;Uid=root;Pwd=test;")
                .UseLoggerFactory(LoggerFactory.Create(builder => builder.AddConsole()));

            return new TestDbContext(optionBuilder.Options);
        }
    }

    public class TestDbContext : DbContext
    {
        public TestDbContext(DbContextOptions<TestDbContext> options) : base(options)
        {
        }

        public DbSet<Test> Tests { get; set; }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            base.OnModelCreating(modelBuilder);

            modelBuilder.Entity<Test>()
                .HasKey(o => new {o.Uid, o.Date});

            modelBuilder.Entity<Test>()
                .Property(o => o.Date)
                .HasDefaultValueSql("CURRENT_TIMESTAMP(6)");
        }
    }

    [Table("test")]
    public class Test
    {
        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        [Column("uid")]
        public int Uid { get; set; }

        [Column("date")]
        public DateTime Date { get; set; }
    }
}
```

Step 3. DB Migration and Update
```
dotnet ef migrations add Init
dotnet ef database update
```

Step 4. Run
```
dotnet run
```

Then occurred exception.
```
info: Microsoft.EntityFrameworkCore.Infrastructure[10403]
      Entity Framework Core 3.1.8 initialized 'TestDbContext' using provider 'Pomelo.EntityFrameworkCore.MySql' with options: None
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (22ms) [Parameters=[@p0='?' (DbType = DateTime)], CommandType='Text', CommandTimeout='30']
      INSERT INTO `test` (`date`)
      VALUES (@p0);
      SELECT `uid`
      FROM `test`
      WHERE ROW_COUNT() = 1 AND `uid` = LAST_INSERT_ID() AND `date` = @p0;
 powerumc@MPOWERUMC-2  ~/Projects/PomeloBug/PomeloBug  dotnet run
info: Microsoft.EntityFrameworkCore.Infrastructure[10403]
      Entity Framework Core 3.1.8 initialized 'TestDbContext' using provider 'Pomelo.EntityFrameworkCore.MySql' with options: None
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (9ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      INSERT INTO `test`
      VALUES ();
      SELECT `uid`, `date`
      FROM `test`
      WHERE ROW_COUNT() = 1 AND `uid` = LAST_INSERT_ID() AND `date` = LAST_INSERT_ID();
Unhandled exception. Microsoft.EntityFrameworkCore.DbUpdateConcurrencyException: Database operation expected to affect 1 row(s) but actually affected 0 row(s). Data may have been modified or deleted since entities were loaded. See http://go.microsoft.com/fwlink/?LinkId=527962 for information on understanding and handling optimistic concurrency exceptions.
   at Microsoft.EntityFrameworkCore.Update.AffectedCountModificationCommandBatch.ThrowAggregateUpdateConcurrencyException(Int32 commandIndex, Int32 expectedRowsAffected, Int32 rowsAffected)
   at Microsoft.EntityFrameworkCore.Update.AffectedCountModificationCommandBatch.ConsumeResultSetWithPropagation(Int32 commandIndex, RelationalDataReader reader)
   at Microsoft.EntityFrameworkCore.Update.AffectedCountModificationCommandBatch.Consume(RelationalDataReader reader)
   at Microsoft.EntityFrameworkCore.Update.ReaderModificationCommandBatch.Execute(IRelationalConnection connection)
   at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.Execute(IEnumerable`1 commandBatches, IRelationalConnection connection)
   at Microsoft.EntityFrameworkCore.Storage.RelationalDatabase.SaveChanges(IList`1 entries)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChanges(IList`1 entriesToSave)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChanges(DbContext _, Boolean acceptAllChangesOnSuccess)
   at Pomelo.EntityFrameworkCore.MySql.Storage.Internal.MySqlExecutionStrategy.Execute[TState,TResult](TState state, Func`3 operation, Func`3 verifySucceeded)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChanges(Boolean acceptAllChangesOnSuccess)
   at Microsoft.EntityFrameworkCore.DbContext.SaveChanges(Boolean acceptAllChangesOnSuccess)
   at Microsoft.EntityFrameworkCore.DbContext.SaveChanges()
   at PomeloBug.Program.Main(String[] args) in /Users/powerumc/Projects/PomeloBug/PomeloBug/Program.cs:line 18
``

**Comment by mguinness** [comment]
Take a look at similar issue https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/issues/801 and try adding `DatabaseGeneratedOption.Identity` to your Date property.

Can you try EF Core 5.0 and [Pomelo 5.0.0-alpha.2](https://www.nuget.org/packages/Pomelo.EntityFrameworkCore.MySql/5.0.0-alpha.2)?  Also try using another EF Core provider to see what occurs.

**Comment by lauxjpn** [comment]
This is an interesting bug, that is basically part of the EF Core default implementation.

The issue is, that key columns that generate values (like auto incremented values or default values) are marked as `read`, so they will be read by the `SELECT` statement that immediately follows the SQL statement.

However, when it comes to generating the `WHERE` clause of this `SELECT` statement, the same batch of columns are first filtered down to only those columns that are part of a key, and are then used by the provider specific method that generates the `WHERE` clause for the `SELECT` statement. These columns can still contain columns that generate values on insert by other means than an auto increment column. This is the case for your `HasDefaultValueSql()` statement.
But EF Core assumes at this point, that all columns left are indeed identity related columns (which is a false assumption) and therefore calls the method, that is responsible for handling only the identity related columns, for all of them.

We fixed this now for Pomelo by filtering out the non-identity related key columns, before letting EF Core handle the rest.

/cc @smitpatel, @roji

**Comment by roji** [comment]
@lauxjpn I think this can be closed as a duplicate of https://github.com/dotnet/efcore/issues/23941

**Comment by lauxjpn** [comment]
> @lauxjpn I think this can be closed as a duplicate of [dotnet/efcore#23941](https://github.com/dotnet/efcore/issues/23941)

@roji @AndriySvyryd Thanks, looks good. I also updated our fix and added a test.
